Upgrade to ORCiD 2.0

thostetler opened the pull request:
**Work in progress**
Functionally complete, at least afaik, I need testers to check things out and make sure that everything is working like it did in prod, and whether there are areas to improve.

This is a big update, most of the changes are in Orcid_api and Extension
* The orcid api actions portion was completely rewritten
  * The getProfile and addWork actions are now debounced to help with number of requests
  * Clean up and better handling of promises throughout, definitely still areas to improve.
* extension
  * Actions portion was completely rewritten
  * Could be optimized more, I think there is information I could gather without making another call.
  * Better error handling, properly rejecting promises.  Hopefully more helpful feedback
    * Have to think about any better ways of providing this feedback to the user, I don't think we want to show a modal with some error message, for example, unless it is helpful in some way to the user.
* Work/Profile
  * Added some classes to help with the parsing/transforming of orcid/ads records
  * Used jsonpath to get/set values using xpath-like paths, which should make maintenance easier in the future.

* Thoughts
  * Do we need to have a database for the api? I think that it really makes more sense when we had to pass the entire profile back to orcid.
    * Could potentially hold a database of items and cache changes, updating the ui to the internal state and pushing up changes after a wait.
  * Edge cases
    * I found a few areas where there were bugs, I thought I had introduced, only to see them show up in prod.  I think in those cases, they are mostly taken care of.  
    * I am not sure how orcid handles the case of multiple sources in 2.0, but I was unclear how to update the `source_name` part for now, until I can find a use case for the changes.

romanchyla commented on src/js/modules/orcid/extension.js:
```diff
@@ -174,123 +177,291 @@ define([
+          var works = profile.getWorks();
+          var matchedWork = _.find(works, function (w) {
+            var wIds = w.getExternalIds();
+            return exIds.bibcode === wIds.bibcode ||
```
bibcode can be only one; but doi and alternate_bibcodes are multivalued - so why check only the first element? ;)

thostetler commented on test/mocha/js/modules/orcid/orcid_extension.spec.js:
```diff
@@ -105,51 +123,75 @@ define([
 
+        get(0, 'update').click();
+        get(1, 'delete').click();
+        get(2, 'add').click();
```
yeah those need to be written out better, the three button clicks are different actions, but right after that it checks to see how many times the merge records method was called (should be once).  There is a lot more that needs to be checked here though, I'd like to test for fails as well.

thostetler commented on test/mocha/js/modules/orcid/work.spec.js:
```diff
@@ -0,0 +1,14 @@
+
+    describe('Instantiation', function () {
+      it('instantiates correctly', function () {
+        expect(new Work({}) instanceof Work).to.equal(true);
```
yeah, I will add in these tests.

thostetler commented on src/js/modules/orcid/orcid_api.js:
```diff
@@ -773,511 +587,348 @@ define([
+          _.forEach(works, function addIdsToDatabase(w, i) {
+            var key;
+            var ids = w.getExternalIds();
+            if (ids.bibcode) {
```
okay yeah that confused me a bit, I'll make that change it should make things a little easier

thostetler commented on src/js/modules/orcid/orcid_api.js:
```diff
@@ -325,404 +288,255 @@ define([
+        }
+
+        var url = this.getUrl('works', putcode);
+        return this.createRequest(url, { method: 'PUT' }, work);
```
Are you saying that the api shouldn't return anything on these requests? or should I be return more generic replies?
```
var promise = $.Deferred();
request(...).done(function (status) {
  if (status === 201) {
    promise.resolve();
  } else {
    promise.reject();
  }
}).fail(function (status) {
  promise.reject();
});
```

romanchyla replied:
im sorry! i wanted to write: "better to *return* promises" universally

thostetler replied:
Ohh! okay yeah that makes sense.  Actually in this case it does do that.  `createRequest` is just a convenience wrapper around the call to `sendData` --> https://github.com/adsabs/bumblebee/pull/1269/files#diff-b88a44bee5c5ab37fb8be1b927cde460R332

There may be a cleaner way of doing it though, I think it makes sense to fail here for status codes that don't make sense than to force the caller to handle the response.

thostetler commented:
@romanchyla this guy is ready for you to check out again.

I fixed, or altered the areas you mentioned.  Some of the tests I've updated.
I was having tons of trouble with the sinon global getting dirtied as it moved through the tests, the state of things would break tests down the line.  for future reference, I was able to eleviate most of that by wrapping all the test code in sinon.test(function () {...});

Thanks!

romanchyla commented:
global variables are usually bad and most likely my ignorance of sinon framework made me use one for the tests. i'm glad it's gone :)

for addWorks failures; if some simple solution is not applicable - such as .done() and chaining of promises, then you could consider using sentinel values; maybe like adding an entry into the cache and removing it after the request finished; if you find some entry there, then you would know the batch failed
